apis/v1alpha1/amazoncloudwatchagent_webhook_test.go

```diff
@@ -0,0 +1,244 @@
+		name     string
+		agent    AmazonCloudWatchAgent
+		expected AmazonCloudWatchAgent
+	}{
+		{
+			name:  "all fields default",
+			agent: AmazonCloudWatchAgent{},
+			expected: AmazonCloudWatchAgent{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						"app.kubernetes.io/managed-by": "amazon-cloudwatch-agent-operator",
+					},
+				},
+				Spec: AmazonCloudWatchAgentSpec{
+					Mode:            ModeDeployment,
+					Replicas:        &one,
+					UpgradeStrategy: UpgradeStrategyAutomatic,
+				},
+			},
+		},
+		{
+			name: "provided values in spec",
+			agent: AmazonCloudWatchAgent{
+				Spec: AmazonCloudWatchAgentSpec{
+					Mode:            ModeSidecar,
+					Replicas:        &five,
+					UpgradeStrategy: "adhoc",
+				},
+			},
+			expected: AmazonCloudWatchAgent{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						"app.kubernetes.io/managed-by": "amazon-cloudwatch-agent-operator",
+					},
+				},
+				Spec: AmazonCloudWatchAgentSpec{
+					Mode:            ModeSidecar,
+					Replicas:        &five,
+					UpgradeStrategy: "adhoc",
+				},
+			},
+		},
+		{
+			name: "Missing route termination",
+			agent: AmazonCloudWatchAgent{
+				Spec: AmazonCloudWatchAgentSpec{
+					Mode: ModeDeployment,
+					Ingress: Ingress{
+						Type: IngressTypeRoute,
+					},
+				},
+			},
+			expected: AmazonCloudWatchAgent{
+				ObjectMeta: metav1.ObjectMeta{
+					Labels: map[string]string{
+						"app.kubernetes.io/managed-by": "amazon-cloudwatch-agent-operator",
+					},
+				},
+				Spec: AmazonCloudWatchAgentSpec{
+					Mode: ModeDeployment,
+					Ingress: Ingress{
+						Type: IngressTypeRoute,
+						Route: OpenShiftRoute{
+							Termination: TLSRouteTerminationTypeEdge,
+						},
+					},
+					Replicas:        &one,
+					UpgradeStrategy: UpgradeStrategyAutomatic,
+				},
+			},
+		},
+	}
+
+	for _, test := range tests {
+		t.Run(test.name, func(t *testing.T) {
+			test.agent.Default()
+			assert.Equal(t, test.expected, test.agent)
+		})
+	}
+}
+
+// TODO: a lot of these tests use .Spec.MaxReplicas and .Spec.MinReplicas. These fields are
+// deprecated and moved to .Spec.Autoscaler. Fine to use these fields to test that old CRD is
+// still supported but should eventually be updated.
+func TestAgentValidatingWebhook(t *testing.T) {
+	three := int32(3)
+
+	tests := []struct { //nolint:govet
+		name        string
+		otelcol     AmazonCloudWatchAgent
+		expectedErr string
+	}{
+		{
+			name:    "valid empty spec",
+			otelcol: AmazonCloudWatchAgent{},
+		},
+		{
+			name: "valid full spec",
+			otelcol: AmazonCloudWatchAgent{
+				Spec: AmazonCloudWatchAgentSpec{
+					Mode:            ModeStatefulSet,
+					Replicas:        &three,
+					UpgradeStrategy: "adhoc",
+					Config: `"agent": {
+   "metrics_collection_interval": 60,
+   "region": "us-west-1",
+   "logfile": "/opt/aws/amazon-cloudwatch-agent/logs/amazon-cloudwatch-agent.log",
+   "debug": false,
+   "run_as_user": "cwagent"
+}
+`,
```

Comment:
nit - the indenting seems off here.

Reply:
This is pretty much taken from https://github.com/open-telemetry/opentelemetry-operator/blob/main/apis/v1alpha1/opentelemetrycollector_webhook_test.go#L174

I think the linter was removed for the config string since yaml indentation matters. But since we use json maybe we don't need to remove the linter. Good catch!